File: tests/python/binary_trees/flatten_binary_tree/test_flatten_binary_tree.py
from problems.binary_trees.assets.BinaryTree import BinaryTree
from problems.binary_trees.flatten_binary_tree.python.flatten_binary_tree import \
    flattenBinaryTree
from tests.python.binary_trees.utility import build_bt
from tests.python.utility import load_test_cases


def are_bts_equal(
    bt1: BinaryTree, bt2: BinaryTree, visited: set[BinaryTree] = None
) -> bool:
    if visited is None:
        visited = set()

    if not bt1 and not bt2:
        return True

    if not bt1 or not bt2 or bt1.value != bt2.value:
        return False

    if bt1 in visited or bt2 in visited:
        return True

    visited.add(bt1)
    visited.add(bt2)

    return are_bts_equal(bt1.left, bt2.left, visited) and are_bts_equal(
        bt1.right, bt2.right, visited
    )


def test_flatten_binary_tree() -> None:
    test_cases = load_test_cases("binary_trees/flatten_binary_tree.json")

    for idx, test in enumerate(test_cases):
        tree = test["tree"]
        expected = test["expected"]

        bt_root = build_bt(tree["nodes"], tree["root"])
        bt_expected = build_bt(expected["nodes"], expected["root"])

        result = flattenBinaryTree(bt_root)

        assert are_bts_equal(result, bt_expected), f"Test case {idx} failed"

File: tests/cpp/binary_trees/test_flatten_binary_tree.cpp
#include "../../../problems/binary_trees/flatten_binary_tree/cpp/flatten_binary_tree.h"
#include "../jsontestbase.h"
#include "./utility.h"

#include <unordered_set>

class FlattenBinaryTreeTest : public JsonTestBase {
public:
  FlattenBinaryTreeTest() {
    json_file_path =
        std::string(TEST_CASES_DIR) + "/binary_trees/flatten_binary_tree.json";
  }
};

bool areBTsEqual(BinaryTree *bt1, BinaryTree *bt2,
                 std::unordered_set<BinaryTree *> &visited) {
  if (!bt1 && !bt2)
    return true;

  if (!bt1 || !bt2 || bt1->value != bt2->value)
    return false;

  if (visited.count(bt1) || visited.count(bt2))
    return true;

  visited.insert(bt1);
  visited.insert(bt2);

  return areBTsEqual(bt1->left, bt2->left, visited) &&
         areBTsEqual(bt1->right, bt2->right, visited);
}

bool areBTsEqual(BinaryTree *bt1, BinaryTree *bt2) {
  std::unordered_set<BinaryTree *> visited;
  return areBTsEqual(bt1, bt2, visited);
}

TEST_F(FlattenBinaryTreeTest, TestCases) {
  ASSERT_FALSE(test_cases.empty()) << "Test cases are empty or not loaded";

  for (const auto &test : test_cases) {
    const auto &treeJson = test["tree"];
    const auto &expectedJson = test["expected"];

    auto root = buildBT(treeJson);
    auto expected = buildBT(expectedJson);

    auto result = flattenBinaryTree(root);

    ASSERT_TRUE(areBTsEqual(result, expected))
        << "Failed for test case: " << test.dump(2);

    deleteBTCycle(root);
    deleteBTCycle(expected);
  }
}
